### PR TITLE
SLCORE-753 Skip project synchronization when branch was recently synchronized

### PR DESCRIPTION
**Pull request by damien-urruty-sonarsource**

### CODE REVIEW

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/sync/SynchronizationService.java [hunk 1]
```diff
@@ -173,15 +178,25 @@ private void synchronizeProjectsOfTheSameConnection(String connectionId, Map<Str
     }
     serverApiProvider.getServerApi(connectionId).ifPresent(serverApi -> {
       var subProgressGap = progressGap / boundScopeBySonarProject.size();
-      var subProgress = progress;
+      var subProgress = new AtomicDouble(progress);
```
Comment (serhat-yenican-sonarsource) [inline backend/core/src/main/java/org/sonarsource/sonarlint/core/sync/SynchronizationService.java:181]: I don't see any concurrent access to the `subProgress` variable. Is there any special reason to use Atomic var?

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/sync/SynchronizationService.java [hunk 2]
```diff
@@ -173,15 +178,25 @@ private void synchronizeProjectsOfTheSameConnection(String connectionId, Map<Str
...
+          .forEach((branchNameOpt, scopes) -> branchNameOpt.ifPresent(branchName -> {
+            var branchBinding = new BranchBinding(new Binding(connectionId, sonarProjectKey), branchName);
+            if (shouldSynchronizeBranch(branchBinding)) {
+              branchSynchronizationTimestampRepository.setLastSynchronizationTimestampToNow(branchBinding);
+              notifier.notify("Synchronizing project '" + sonarProjectKey + "'...", (int) Math.round(subProgress.get()));
+              issueSynchronizationService.syncServerIssuesForProject(serverApi, connectionId, sonarProjectKey, branchName, cancelMonitor);
+              taintSynchronizationService.synchronizeTaintVulnerabilities(serverApi, connectionId, sonarProjectKey, branchName, cancelMonitor);
+              hotspotSynchronizationService.syncServerHotspotsForProject(serverApi, connectionId, sonarProjectKey, branchName, cancelMonitor);
+              synchronizedConfScopeIds.addAll(entry.getValue().stream().map(BoundScope::getConfigScopeId).collect(toSet()));
+              subProgress.getAndAccumulate(subProgressGap, Double::sum);
```
Comment (serhat-yenican-sonarsource) [inline backend/core/src/main/java/org/sonarsource/sonarlint/core/sync/SynchronizationService.java:197]: I think the accumulation of subProgress should be right under the for loop that iterates the `boundScopeBySonarProject` collection, and `subProgress` was divided into its size.

**Review by serhat-yenican-sonarsource — COMMENTED** [review]
Looks good!
Just two minor comments

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/sync/SynchronizationService.java (hunk 1 above)
Comment (damien-urruty-sonarsource) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/sync/SynchronizationService.java:181]: The variable is used in a lambda, so I could not modify it directly and was getting a compilation error:

> Variable used in lambda expression should be final or effectively final

The other technique I encountered would be to use an array of 1 element, and to modify it everytime

If you see a better approach, let me know

Comment (serhat-yenican-sonarsource) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/sync/SynchronizationService.java:181]: Ahh, lambda, true. I didn't think that. I think your approach is quite reasonable.

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/sync/SynchronizationService.java (hunk 2 above)
Comment (damien-urruty-sonarsource) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/sync/SynchronizationService.java:197]: I think you're right. It was already wrong before, I will fix this

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/sync/SynchronizationService.java (hunk 1 above)
Comment (damien-urruty-sonarsource) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/sync/SynchronizationService.java:181]: I ended up extracting the lambda to a separate method

**Review by serhat-yenican-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**